include/dlaf/eigensolver/get_band_size.h

```diff
@@ -0,0 +1,35 @@
+//
+// Distributed Linear Algebra with Future (DLAF)
+//
+// Copyright (c) 2018-2022, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+#pragma once
+
+#include "dlaf/eigensolver/eigensolver/api.h"
+
+#include "dlaf/common/assert.h"
+#include "dlaf/tune.h"
+#include "dlaf/types.h"
+
+namespace dlaf::eigensolver::internal {
+
+// Returns the smallest divisor of nb larger than b_min = getTuneParameters().eigensolver_min_band.
+// If nb is smaller than b_min returns nb.
+inline SizeType getBandSize(SizeType nb) noexcept {
+  SizeType b_min = getTuneParameters().eigensolver_min_band;
+
+  DLAF_ASSERT(nb >= 1, nb);
+  DLAF_ASSERT(b_min >= 2, b_min);
+
+  for (SizeType div = nb / b_min; div > 1; --div) {
+    if (nb % div == 0)
+      return nb / div;
+  }
```

Comment:
This looks correct to me.

I just want to link https://en.cppreference.com/w/cpp/numeric/math/div since I learned about it now while looking at this. It combines the modulo and division into one operation. From what I understand the current `%` then `/` is almost guaranteed to be optimized to do the same so I don't think you need to change it unless you would find it clearer.

Finally, I was trying to understand why you've chosen to loop "backwards" over `div`, as opposed to "forwards" from `b_min`. Is the reason that in typical cases (`b_min` is a large fraction of `nb`) you'll end up doing fewer iterations? I think there are pathological cases in either implementation, but this seems like a reasonable choice in this case if I understand this correctly.

Reply:
> This looks correct to me.
> 
> I just want to link https://en.cppreference.com/w/cpp/numeric/math/div since I learned about it now while looking at this. It combines the modulo and division into one operation. From what I understand the current `%` then `/` is almost guaranteed to be optimized to do the same so I don't think you need to change it unless you would find it clearer.

Good to know! However I would trust the compiler here. (Even if the optimization is not made the integer division is performed just one extra time).

> Finally, I was trying to understand why you've chosen to loop "backwards" over `div`, as opposed to "forwards" from `b_min`. Is the reason that in typical cases (`b_min` is a large fraction of `nb`) you'll end up doing fewer iterations? I think there are pathological cases in either implementation, but this seems like a reasonable choice in this case if I understand this correctly.

Exactly, the block size should be 256-1024 (from benchmarks) and the band size absolute lower bound is 64 (likely the best choice is around 100), therefore is better to iterate over div.

Reply:
Thanks for the explanation! Agreed on not caring about the (possible) optimization. The whole function call is anyway likely an insignifcant rounding error compared to all the other work for the eigensolver...